refined/src/main/scala/magnolify/refined/package.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package magnolify
+
+import eu.timepit.refined.api.{RefType, Validate}
+
+package object refined extends RefinedAvro with RefinedBigQuery with RefinedBigTable
+
+trait RefinedBase extends Serializable {
+  def wrap[F[_, _], T, P](t: T)(implicit rt: RefType[F], v: Validate[T, P]): F[T, P] = {
+    val res = v.validate(t)
+    if (res.isPassed)
+      rt.unsafeWrap[T, P](t)
+    else
+      throw new RuntimeException(
+        s"Refined predicate Failed: ${v.showExpr(t)}"
+      )
+  }
+}
+
+trait RefinedAvro extends RefinedBase {
+  import magnolify.avro._
+  import eu.timepit.refined.api.{RefType, Validate}
+
+  implicit def refAvroT[F[_, _], T, P](implicit
+    af: AvroField[T],
+    rt: RefType[F],
+    v: Validate[T, P]
+  ): AvroField[F[T, P]] =
+    AvroField.from[T](wrap[F, T, P])(rt.unwrap)
```

Comment:
Are these wrapping/unwrapping the reason we can't support some other types like `Entity`/`Example`/`Protobuf`? And we have to inject validation logic on a leaf field level, and can't do it just for the case class level?